### DISCUSSION TITLE
Laravel Horizon Default Supervisor Inconsistently Going Offline

### DISCUSSION THREAD

**Discussion by hyp-azam** [laravel/framework#49299] [Q&A] [answered]
## Laravel Horizon Supervisor Inconsistently Going Offline

**Versions:**
- Laravel version: v10.35.0
- Horizon version: v5.21.4

**Environment:**
Running horizon on an ECS container

**Problem:**
I recently upgraded Laravel from 8 to 9 and then finally 10. My horizon command is working and jobs are processing but on the dashboard, one of my supervisors (a default supervisor "supervisor-1") intermittently goes offline and then comes online again. 

Upon investigating, I checked the horizon routes specifically `{{host}}/horizon/api/masters` API which has the supervisors array. There I saw that `supervisor-1` was in running state with some id and then after some time when another API was called the supervisor was in an inactive state. Then after some time on another API, `supervisor-1` was in running state but the PID had changed from what it was earlier. This indicates that it was somehow restarted. As for other supervisors which are not having the problem, the PID remains the same. 

I have checked supervisor logs and ECS container logs but did not find anything unusual.

1. supervisor-1 running with PID: 20171

![image](https://github.com/laravel/framework/assets/115896215/c2895c39-8820-44cf-8944-1ff757f8d373)

2. supervisor-1 now inactive

![image](https://github.com/laravel/framework/assets/115896215/3868311a-3f3b-4b3e-8f96-2a1984ccb460)

3. supervisor-1 running again with PID: 20517 (changed)

![image](https://github.com/laravel/framework/assets/115896215/f5ab3848-f566-48cc-9bc5-178567206e58)

**Supervisor Configuration:**
```php
'defaults' => [
    'supervisor-1' => [
        'connection' => 'redis',
        'balance' => 'simple',
        'queue' => ['default'],
        'maxTime' => 3600,
        'maxJobs' => 1000,
        'minProcesses' => 5,
        'maxProcesses' => 10,
        'memory' => 32,
        'tries' => 1,
        'balanceMaxShift' => 5,
        'balanceCooldown' => 3
    ],
],

**Comment by hyp-azam** [comment, marked as answer]
I solved the issue. The problem was caused by the memory allocation for supervisor-1, which was set to 32MB, while the other supervisors had 64MB allocated. When I increased the memory to 64MB, it worked fine. I am just surprised that Laravel did not provide any logs regarding the supervisor crashing in the Horizon logs.

**Reply by siddiquiazam** [reply]
It seems we can listen to the following events provided by horizon:
1. MasterSupervisorOutOfMemory
2. SupervisorOutOfMemory

We can add listeners for these events and then log or alert however we want.
